crates/oxc_regexp_parser/src/parser/body_parser/reader.rs

```diff
@@ -0,0 +1,226 @@
+pub struct Reader<'a> {
+    source: &'a str,
+    char_indices: std::str::CharIndices<'a>,
+    unicode_mode: bool,
+    index: usize,
+}
+
+impl<'a> Reader<'a> {
+    pub fn new(source: &'a str, unicode_mode: bool) -> Self {
+        Self { source, char_indices: source.char_indices(), unicode_mode, index: 0 }
+    }
+
+    // NOTE: Should be decoupled from the reader...?
+    // ```
+    // let pos = Position::new(source, unicode_mode);
+    // pos.get(reader.index);
+    // ````
+    pub fn position(&self) -> usize {
+        let mut char_indices = self.char_indices.clone();
+
+        if self.unicode_mode {
+            char_indices.nth(self.index).map_or(self.source.len(), |(i, _)| i)
+        } else {
+            let mut utf16_units = 0;
+            let mut byte_index = 0;
+            for (idx, ch) in char_indices {
+                if utf16_units == self.index {
+                    return idx;
+                }
+
+                utf16_units += ch.len_utf16();
+                byte_index = idx + ch.len_utf8();
+            }
+            byte_index
+        }
+    }
+
+    pub fn rewind(&mut self, index: usize) {
+        self.index = index;
+    }
+
+    pub fn advance(&mut self) {
+        self.index += 1;
+    }
+
+    fn peek_nth(&self, n: usize) -> Option<u32> {
+        let nth = self.index + n;
+
+        // NOTE: This may affect performance?
```

Comment:
Yes, `.chars().next()` is really expensive.

Take a look at `lookahead` in `crates/oxc_parser/src/lexer/mod.rs.